Big Tree Tech EBB42 v1.1 board support

alextrical opened the pull request:
<!--

Submitting a Pull Request

- Please fill out all sections of this form. You can delete the helpful comments.
- Pull Requests without clear information will take longer and may even be rejected.
- We get a high volume of submissions so please be patient during review.

-->

### Description

<!--

Clearly describe the submitted changes with lots of details. Include images where helpful. Initial reviewers may not be familiar with the subject, so be as thorough as possible. You can use MarkDown syntax to improve readability with bullet lists, code blocks, and so on. PREVIEW and fix up formatting before submitting.

-->
Add basic support for Big Tree Tech [EBB42 V1.1](https://github.com/bigtreetech/EBB/tree/master/EBB%20CAN%20V1.1%20(STM32G0B1)/EBB42%20CAN%20V1.1)
This board is ideal for use will Filament Extruder/Pulstruder systems.
Mainly under development for the [Polyformer](https://github.com/Reiten966/Polyformer) project, though this will be of use to numerous other Filament making projects, that use Printer control boards currently.
Such as [The Recreator](https://hackaday.io/project/185867-the-recreator-3d-make-recycled-pet1-3d-filament) that uses Marlin compiled for the Ender 3

### Requirements

<!-- Does this PR require a specific board, LCD, etc.? -->
A Big Tree Tech EBB42 V1.1

### Benefits

<!-- What does this PR fix or improve? -->
Board is now supported

### Configurations

<!-- Attach Configurations ZIP and any other files needed to test this PR. -->
#define MOTHERBOARD BOARD_BTT_EBB42_V1_1
#define SERIAL_PORT -1

### Related Issues

<!-- Does this PR fix a bug or fulfill a Feature Request? Link related Issues here. -->
Currently compiles, can run a SSD1306 display over I2C, responds to serial requests over USB. Other functionality needs testing.

alextrical committed:
Initial support for BTT EBB42 V1.1

alextrical committed:
Reverted config file back to default

removed accidental commit of Configuration.h

thisiskeithb commented:
What’s the use case for this board under Marlin? Seems like it’s missing most basic I/O for a 3D printer motherboard.

ellensp commented:
Is being used for a filament extruder controller.  basically E axis + hotend + i2c display only
But he wanted to run Marlin on it not kipper, so a worthy goal. (in my view) 
No its not for 3d printer.

thisiskeithb commented:
I think it should be made more clear (through comments + environment rename like `*_filament_extruder`) that this board is not a general 3D printing controller or supported as a toolboard via CANBUS (as it was originally designed) since I can already envision the potential questions about this being supported on a standard machine.

Does Marlin compile with only an E-axis?

You'll probably also want to remove all the unused pins / features in the pins file for clarity.

alextrical commented:
> What’s the use case for this board under Marlin? Seems like it’s missing most basic I/O for a 3D printer motherboard.

I'm trying to use this for a Filament Pullstrusion/Extrusion system, where all that is required is an Extruder, with stepper motor and PID control. https://github.com/Reiten966/Polyformer

I'm hoping once I get the basic functionality of Marlin running, to work on maybe getting marlin to support 0/None kinematics, have a specific landing screen for Extrusion only devices (such as plastic to filament machines), and work on a internal GCode loop build into the firmware to run the motor and heater loop until stopped from the menu.

I know this is a niche use case, but there are numerous different individuals building such systems from scratch, using Kliper, or using marlin with a SD card and Gcode, it just currently feels klunky, or re-treading the work that Marlin has already tried and tested. PID tuning for example is a pain to do correctly, and the decoupling safety checks done here are nice.

It would also be nice to add this functionality to the core repo for all to benefit from, rather than having a dedicated fork for the more basic system for Filament extruders/pullstruders

alextrical commented:
> I think it should be made more clear (through comments + environment rename like *_filament_extruder)
Can do, I will make that change if that will help get this board officially supported, and able to be easily used by the community

> Does Marlin compile with only an E-axis?
Unfortunately not, it triggers the sanity checks for missing X, Y & Z. I will be looking into how to disable that requirement, and then request advice on how to make a Define to allow Extruder only functionality to compile without triggering the warnings

> You'll probably also want to remove all the unused pins / features in the pins file for clarity.
Can do.

alextrical committed:
Ammended pins and build flags to set I2C pins

alextrical committed:
amended env name and warnings

Added warning to ENV to warn users the board is not a usual 3d printer board, but intended to be used as a plastic to filament extruder, and wont work as a printer or canbus control board

alextrical commented:
Added warnings about the board not being a 3D printer control or a Hot end Canbus board to the env file, and added the recommended extension to the environment name to help reduce the chance of future user confusion about the board capabilities.

Superfluous pins have been removed from the Pins file, however pins for XYZ remain to allow Marlin to compile (As there is no option for None Kinematics yet)

alextrical committed:
Merge branch 'MarlinFirmware:bugfix-2.1.x' into BTT_EBB42_V1.1

alextrical committed:
Corrected typo in E0 name

replaced E1 with the correct value of E0, so that USART connection to the TMC2209 works as expected

alextrical committed:
Merge branch 'MarlinFirmware:bugfix-2.1.x' into BTT_EBB42_V1.1

alextrical committed:
Added EEPROM config

alextrical committed:
Merge branch 'BTT_EBB42_V1.1' of https://github.com/alextrical/Marlin into BTT_EBB42_V1.1

alextrical committed:
EEPROM Bug fix

alextrical commented:
Virtual EEPROM now works

thisiskeithb commented on Marlin/src/pins/stm32g0/pins_BTT_EBB42_V1_1.h:
```diff
@@ -0,0 +1,157 @@
+    #define FLASH_EEPROM_EMULATION
+  #endif
+  #define EEPROM_PAGE_SIZE      (0x800UL) // 2K
+  #define EEPROM_START_ADDRESS      (0x8000000UL + (FLASH_SIZE) - (EEPROM_PAGE_SIZE))
```
`EEPROM_START_ADDRESS` is usually defined as the following in other STM32 pins files:
```cpp
#define EEPROM_START_ADDRESS (0x8000000UL + (STM32_FLASH_SIZE) * 1024UL - (EEPROM_PAGE_SIZE) * 2UL)
```

thinkyhead committed:
renumber boards

thinkyhead committed:
cleanup

alextrical committed:
EEPROM Fix and defined *_STOP_PIN to allow compile to complete

EEPROM calculation incorrectly evaluating to the required start address for page 63 of the flash
*_STOP_PIN's need to be defined to a value to allow marlin to compile. Future revisions will be looked into to remove the dependency of XY kinematics on machines that only require extruders

alextrical committed:
Merge branch 'MarlinFirmware:bugfix-2.1.x' into BTT_EBB42_V1.1

alextrical committed:
Merge branch 'MarlinFirmware:bugfix-2.1.x' into BTT_EBB42_V1.1

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/24964

thinkyhead committed:
Merge 'bugfix-2.1.x' into pr/24964

thinkyhead committed:
cleanup

thinkyhead committed:
Merge 'bugfix-2.1.x' into pr/24964

thinkyhead committed:
etc
